src/main/scala/com/databricks/labs/overwatch/pipeline/SilverTransforms.scala

```diff
@@ -1098,23 +1099,50 @@ trait SilverTransforms extends SparkSessionWrapper {
       .withColumn("target_num_workers", last('target_num_workers, true).over(stateUnboundW))
       .withColumn("current_num_workers", last('current_num_workers, true).over(stateUnboundW))
       .withColumn("unixTimeMS_state_start", 'timestamp)
-      .withColumn("unixTimeMS_state_end", coalesce( // if state end open, use pipelineSnapTime, will be merged when state end is received
-        lead('timestamp, 1).over(stateUnboundW) - lit(1), // subtract 1 millis
-        lit(untilTime.asUnixTimeMilli)
-      ))
+      //Change for Issue-846
+
+    val clusterBaseDF = clusterBase(auditLogDF)
+    val clustersRemoved = clusterBaseDF
+      .filter('actionName.isin("permanentDelete"))
+      .select('organization_id, 'cluster_id, 'userEmail.alias("deleted_by"))
+
+    val removedClusterID = clustersRemoved.select("cluster_id").distinct().collect().map(f=>f(0))
+    val w = Window.partitionBy('organization_id, 'cluster_id).orderBy('unixTimeMS_state_start)
+    val timeStampForRemovedCluster = clusterEventsBaselineFinal
+      .drop("driverSpecs", "workerSpecs")
+      .filter(col("cluster_id").isin(removedClusterID: _*))
+      .withColumn("positiveTestCase", when(last('state).over(w) === "RUNNING", lit(true)).otherwise(lit(false)))
+      .withColumn("positiveTestCase2", lead('positiveTestCase,1).over(w))
+      .filter('positiveTestCase2.isNull)
+      .filter('state === "RUNNING").select("timestamp").collect().map(t=>t(0))
+
+    clusterEventsBaselineFinal
+      .withColumn("state",when(col("timestamp").isin(timeStampForRemovedCluster: _*),lit("PERMANENT_DELETE")).otherwise('state))
+      .withColumn("isRunning",when(col("timestamp").isin(timeStampForRemovedCluster: _*),lit(false)).otherwise('isRunning))
+      .withColumn("unixTimeMS_state_start",when(col("timestamp").isin(timeStampForRemovedCluster: _*),lit(System.currentTimeMillis)).otherwise('unixTimeMS_state_start))
+      .withColumn("unixTimeMS_state_end",when(col("timestamp").isin(timeStampForRemovedCluster: _*),lit(System.currentTimeMillis))
```

Comment:
this logic doesn't make sense. I think you should do this differently. Move it to the end identify your billable running clusters and join on cluster_id and keep the timestamp for the deleted fields and make that the timestamp field.

Reply:
Done